### PR TITLE
test: staking contract

### LINKED ISSUE #2436: Staking pool contract. Nightly test
Implement a nightly test for staking & delegation contract https://github.com/near/staking-contract/.

Scenario:
 - Start a singe node
 - Deploy staking contract
 - Start another node that uses this staking contract.
 - Delegate funds enough to stake
 - Validate that second node is staking 
 - Undelegate funds
 - Second node stops staking
 - Funds received back by delegator contain correct amount including rewards.

### PR DESCRIPTION
**Pull request by ilblackdragon**
Test staking contract in nightly infrastructure. Fixes #2436 

Additionally adds:
 - Account & JsonProvider abstractions, similar to near-api-js. This makes subset of pytest/lib really a near-api-py library at this point.
 - Cluster abstract to manage cluster.

Test plan
----
nightly passes with new staking_contract.py test

### CODE REVIEW

File: pytest/lib/account.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+class JsonProvider(object):
+
+    def __init__(self, rpc_addr):
+        self._rpc_addr = rpc_addr
+
+    def rpc_addr(self):
+        return self._rpc_addr
+
+    def json_rpc(self, method, params, timeout=2):
```
Comment (SkidanovAlex) [inline pytest/lib/account.py:23]: This logic seems to come from `LocalNode`. Probably should be removed there?

**Comment by evgenykuzyakov** [comment]
Added staking pool reward fee. Here is the latest log:
```
pytest  python3 tests/contracts/staking_contract.py                      1.4m  Tue May  5 16:37:30 2020
Use default config {'local': True, 'near_root': '../target/debug/', 'binary_name': 'neard'}
Creating LOCAL cluster configuration with 1 nodes
Search for stdout and stderr in ['/Users/ekwork/.near/test0']
Starting node 0 as BOOT NODE
node 0 started
Starting node 1 with boot=ed25519:He7QeRuwizNEhBioYG3u4DZ8jWXyETiyNzFD3MkTjDMf@127.0.0.1:24577
node 1 started
{'amount': '100000000000425342547765000', 'locked': '0', 'code_hash': 'DpywzzTNLvpUExfMg8x6yUPuNzEXjQ22D5MpnpKxQunK', 'storage_usage': 232109, 'storage_paid_at': 0, 'block_height': 11, 'block_hash': 'DYfY11zyT8bK8wPNpDuDY3ZFyU5YAQuJzwVVod3sRKLJ'}
>>> @user1 deposit call
Log: @user1 deposited 50000000000000000000000000000000. New unstaked balance is 50000000000000000000000000000000
>>> @user1 unstaked balance = 50000000000000000000000000000000
>>> @user1 stake call
Log: Epoch 2: Contract received total rewards of 818315268425000 tokens. New total staked balance is 818315268425000. Total number of shares 818315268425000
Log: Total rewards fee is 818315268425000 tokens or 818315268425000 stake shares.
Log: @user1 staking 50000000000000000000000000000000. Received 50000000000000000000000000000000 new staking shares. Total 0 unstaked balance and 50000000000000000000000000000000 staking shares
Log: Contract total staked balance is 50000000000000000818315268425000. Total number of shares 50000000000000000818315268425000
>>> @user1 stake = 50000000000000000000000000000000
>>> Waiting for 20 blocks
Log: Epoch 3: Contract received total rewards of 816373399365000 tokens. New total staked balance is 50000000000000001634688667790000. Total number of shares 50000000000000000899952608361500
Log: Total rewards fee is 81637339936500 tokens or 81637339936500 stake shares.
Log: Epoch 4: Contract received total rewards of 1511182480385000 tokens. New total staked balance is 50000000000000003145871148175000. Total number of shares 50000000000000001051070856400000
Log: Total rewards fee is 151118248038500 tokens or 151118248038500 stake shares.
{'current_validators': [{'account_id': 'staker', 'public_key': 'ed25519:7iPo7pPULX2ceZKC8Pf6Zy8h8DBq752ehgKB7ZNJoCU2', 'is_slashed': False, 'stake': '50000000000000000818315268425000', 'shards': [0], 'num_produced_blocks': 2, 'num_expected_blocks': 2}, {'account_id': 'test0', 'public_key': 'ed25519:He7QeRuwizNEhBioYG3u4DZ8jWXyETiyNzFD3MkTjDMf', 'is_slashed': False, 'stake': '50000112728313217162578465185838', 'shards': [0], 'num_produced_blocks': 0, 'num_expected_blocks': 0}], 'next_validators': [{'account_id': 'staker', 'public_key': 'ed25519:7iPo7pPULX2ceZKC8Pf6Zy8h8DBq752ehgKB7ZNJoCU2', 'stake': '50000014269395285139804934380876', 'shards': [0]}, {'account_id': 'test0', 'public_key': 'ed25519:He7QeRuwizNEhBioYG3u4DZ8jWXyETiyNzFD3MkTjDMf', 'stake': '50000126997732528148889511605794', 'shards': [0]}], 'current_fishermen': [], 'next_fishermen': [], 'current_proposals': [], 'prev_epoch_kickout': [], 'epoch_start_height': 51}
>>> @user2 deposit call
Log: Epoch 5: Contract received total rewards of 14269395285833248038170876 tokens. New total staked balance is 50000014269395288979119186345876. Total number of shares 50000001426939163126334455657037
Log: Total rewards fee is 1426939528583324803817087 tokens or 1426939162075263599257037 stake shares.
Log: @user2 deposited 70000000000000000000000000000000. New unstaked balance is 70000000000000000000000000000000
>>> @user2 stake call
Log: @user2 staking 70000000000000000000000000000000. Received 69999982020566554918959372095949 new staking shares. Total 0 unstaked balance and 69999982020566554918959372095949 staking shares
Log: Contract total staked balance is 120000014269395288979119186345876. Total number of shares 119999983447505718045293827752986
Log: Epoch 6: Contract received total rewards of 14269391446190177120870015 tokens. New total staked balance is 120000028538786735169296307215891. Total number of shares 119999984874444343444773498976323
Log: Total rewards fee is 1426939144619017712087001 tokens or 1426938625399479671223337 stake shares.
>>> @user1 stake = 50000018193478289772454989120360
>>> @user1 unstake call
Log: @user1 unstaking 50000018193478289772454989120360. Spent 50000000000000000000000000000000 staking shares. Total 50000018193478289772454989120360 unstaked balance and 0 staking shares
Log: Contract total staked balance is 70000010345308445396841318095531. Total number of shares 69999984874444343444773498976323
>>> @user1 stake (after unstake) = 0
Log: Epoch 7: Contract received total rewards of 14269391665042398657569343 tokens. New total staked balance is 70000024614700110439239975664874. Total number of shares 69999986301382728938242707225599
Log: Total rewards fee is 1426939166504239865756934 tokens or 1426938385493469208249276 stake shares.
Log: Epoch 8: Contract received total rewards of 14269391892434129452945495 tokens. New total staked balance is 70000038884092002873369428610369. Total number of shares 69999987728320875379740947066455
Log: Total rewards fee is 1426939189243412945294549 tokens or 1426938146441498239840856 stake shares.
Log: Epoch 9: Contract received total rewards of 16647623766019483056296095 tokens. New total staked balance is 70000055531715768892852484906464. Total number of shares 69999989393081679052798893143212
Log: Total rewards fee is 1664762376601948305629609 tokens or 1664760803673057946076757 stake shares.
{'current_validators': [{'account_id': 'staker', 'public_key': 'ed25519:7iPo7pPULX2ceZKC8Pf6Zy8h8DBq752ehgKB7ZNJoCU2', 'is_slashed': False, 'stake': '70000038884092000617507373035369', 'shards': [0], 'num_produced_blocks': 4, 'num_expected_blocks': 4}, {'account_id': 'test0', 'public_key': 'ed25519:He7QeRuwizNEhBioYG3u4DZ8jWXyETiyNzFD3MkTjDMf', 'is_slashed': False, 'stake': '50000169806004035481677478725620', 'shards': [0], 'num_produced_blocks': 4, 'num_expected_blocks': 4}], 'next_validators': [{'account_id': 'staker', 'public_key': 'ed25519:7iPo7pPULX2ceZKC8Pf6Zy8h8DBq752ehgKB7ZNJoCU2', 'stake': '70000055531715765197565403246464', 'shards': [0]}, {'account_id': 'test0', 'public_key': 'ed25519:He7QeRuwizNEhBioYG3u4DZ8jWXyETiyNzFD3MkTjDMf', 'stake': '50000181697196676129207919048282', 'shards': [0]}], 'current_fishermen': [], 'next_fishermen': [], 'current_proposals': [], 'prev_epoch_kickout': [], 'epoch_start_height': 91}
>>> Waiting for another 20 blocks
>>> @user1 stake = 0
>>> @user1 unstaked balance = 50000018193478289772454989120360
>>> @user1 withdraw call
Log: Epoch 10: Contract received total rewards of 16647623464632931784215815 tokens. New total staked balance is 70000072179339233525784269122279. Total number of shares 69999991057842096260739636248730
Log: Total rewards fee is 1664762346463293178421581 tokens or 1664760417207940743105518 stake shares.
Log: @user1 withdrawing 50000018193478289772454989120360. New unstaked balance is 0
>>> @user1 unstaked balance = 0
>>> @pool_owner unstaked balance = 0
>>> @pool_owner stake = 9037286014447717518433748
Cleaning up node 127.0.0.1:24578 on script exit
Executed refmap tests: 0
Cleaning up node 127.0.0.1:24577 on script exit
Executed refmap tests: 0
```

**Comment by ilblackdragon** [comment]
Note, added a bounty to pull the Python library out of this PR: https://github.com/near/community/issues/11

File: pytest/lib/cluster.py [hunk 2]
```diff
@@ -73,7 +76,7 @@ class BaseNode(object):
     def _get_command_line(self, near_root, node_dir, boot_key, boot_node_addr, binary_name='neard'):
         if boot_key is None:
             assert boot_node_addr is None
-            return [os.path.join(near_root, binary_name), "--verbose", "", "--home", node_dir, "run"]
+            return [os.path.join(near_root, binary_name), "--verbose", "", "--home", node_dir, "run", "--archive"]
```
Comment (SkidanovAlex) [inline pytest/lib/cluster.py:79]: We don't want to default to `archive`, better to make it a flag to the method.

File: pytest/tests/contracts/staking_contract.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Spins up one node, deploy staking contract, spin up another node that stakes with this staking contract.
```
Comment (SkidanovAlex) [inline pytest/tests/contracts/staking_contract.py:1]: You should also add it to `nightly/nightly.txt
`

**Comment by ilblackdragon** [comment]
Tooling went leaps forward. 
Closing this.
If there is anything interesting here for nightly testing @bowenwang1996 - please salvage.

**Pull request closed without merging**